File: chunky_proxy/src/clusters.js
const Chunks = require('./chunks');
const Util = require('./util');

class Clusters {
    constructor() {
        this.chunkState = new Chunks();
        this.blobs = [];
    }

    updateLoaded(server, chunks) {
        // dict of stringify(chunk) -> chunk
        const oldChunks = this.chunkState.getServerLoaded(server);
        const newChunks = {}
        for (let chunk of chunks) {
            newChunks[chunk] = chunk;
        }

        const unloadedChunks = [];
        const newlyLoadedChunks = [];
        for (let chunk in newChunks) {
            if (!(chunk in oldChunks)) {
                newlyLoadedChunks.push(newChunks[chunk]);
            }
        }
        for (let chunk in oldChunks) {
            if (!(chunk in newChunks)) {
                unloadedChunks.push(oldChunks[chunk]);
            }
        }

        for (let chunk of newlyLoadedChunks) {
            this.chunkState.addLoaded(chunk[0], chunk[1], server);
        }
        for (let chunk of unloadedChunks) {
            this.chunkState.removeLoaded(chunk[0], chunk[1], server);
        }

        const allLoadedChunks = this.chunkState.getAllLoadedChunks();
        const visited = {};
        const out = [];

        const dfs = function (chunk, blob) {
            if (chunk in visited) return;
            if (!(chunk in allLoadedChunks)) return;
            visited[chunk] = true;
            blob[chunk] = chunk;
            for (let neighboring of Util.getNeighboring(chunk)) {
                dfs(neighboring, blob);
            }
        };

        for (let chunkKey of Object.keys(allLoadedChunks)) {
            if (chunkKey in visited) {
                continue
            }
            const chunk = allLoadedChunks[chunkKey];
            const blob = {};
            dfs(chunk, blob);
            
            out.push(blob);
        }
        this.blobs = out;
    }

    redistribute(serverList) {
        const serverAssignments = {};
        for (let server of serverList) {
            serverAssignments[server] = {
                numChunks: 0,
                blobs: [],
            };
        }
        for (let blob of this.blobs) {
            const serverCounts = {}
            let maxServer = null;
            let maxCount = -1;
            const blobChunks = Object.values(blob)
            for (let chunk of blobChunks) {
                const servers = this.chunkState.getLoaded(chunk[0], chunk[1]);
                for (let server in servers) {
                    if (!(server in serverCounts)) {
                        serverCounts[server] = 0;
                    }
                    serverCounts[server]++;
                    if (serverCounts[server] > maxCount) {
                        maxServer = server;
                        maxCount = serverCounts[server];
                    }
                }
            }
            serverAssignments[maxServer].numChunks += blobChunks.length;
            serverAssignments[maxServer].blobs.push(blob);
        }
        if (serverList.length == 1) {
            return serverAssignments;
        }
        while (true) {
            let minChunkServer = null;
            let minChunkNumber = Number.MAX_SAFE_INTEGER;

            let maxChunkServer = null;
            let maxChunkNumber = -1;
            let maxChunkLastBlobSize = null;
            for (let server in serverAssignments) {
                const assignment = serverAssignments[server];
                if (assignment.numChunks < minChunkNumber) {
                    minChunkServer = server;
                    minChunkNumber = assignment.numChunks;
                }
                if (assignment.numChunks > maxChunkNumber) {
                    if (assignment.blobs.length == 0) {
                        continue
                    }
                    maxChunkServer = server;
                    maxChunkNumber = assignment.numChunks;
                    maxChunkLastBlobSize = Object.keys(assignment.blobs[assignment.blobs.length - 1]).length;
                }
            }
            if (maxChunkNumber - maxChunkLastBlobSize < minChunkNumber + maxChunkLastBlobSize) {
                break;
            }
            serverAssignments[minChunkServer].blobs.push(serverAssignments[maxChunkServer].blobs.pop());
        }
        return serverAssignments;
    }

    getTransfers(serverList) {
        const newServerAssignments = this.redistribute(serverList);

        const transfers = {};
        for (let server in newServerAssignments) {
            const assignment = newServerAssignments[server];
            for (let blob of assignment.blobs) {
                for (let chunkKey in blob) {
                    const chunk = blob[chunkKey];
                    if (this.chunkState.isOwned(chunk[0], chunk[1])) {
                        const previousOwner = this.chunkState.getOwner(chunk[0], chunk[1]);
                        if (previousOwner == server) {
                            // do nothing, we already own it
                        } else {
                            // we don't own the chunk, transfer ownership
                            console.log(chunkKey + ' owned by ' + previousOwner);
                            const key = {
                                from: previousOwner,
                                to: server,
                            };
                            if (!(key in transfers)) {
                                transfers[key] = {
                                    from: key.from,
                                    to: key.to,
                                    chunks: [],
                                };
                            }
                            transfers[key].chunks.push(chunk);
                        }
                    } else {
                        // nobody owns it - we need to set ownership
                        this.chunkState.setOwner(chunk[0], chunk[1], server);
                    }
                }
            }
        }
        return transfers;
    }
}

module.exports = Clusters;

File: chunky_manager/go/src/github.com/omegablitz/chunky_manager/clusters.go
package main

import (
	"math"
	"sync"
)

type clusters struct {
	sync.Mutex
	chunkState *chunkManager
	blobs      []map[Chunk]struct{}
}

func NewClusters() *clusters {
	return &clusters{
		chunkState: NewChunkManger(),
	}
}

func (cl *clusters) UpdateLoadedChunks(server Server, chunks []Chunk) {
	cl.Lock()
	defer cl.Unlock()

	oldLoadedChunks := cl.chunkState.getServerLoadedChunks(server)
	currentLoadedChunks := make(map[Chunk]struct{})
	for _, chunk := range chunks {
		currentLoadedChunks[chunk] = struct{}{}
	}

	newLoadedChunks, unloadedChunks := chunkChanges(oldLoadedChunks, currentLoadedChunks)
	for chunk := range newLoadedChunks {
		cl.chunkState.addLoadedServer(chunk, server)
	}
	for chunk := range unloadedChunks {
		cl.chunkState.removeLoadedServer(chunk, server)
	}

	var blobs []map[Chunk]struct{}
	allLoadedChunks := cl.chunkState.getAllLoadedChunks()

	visited := make(map[Chunk]struct{})
	var dfs func(chunk Chunk, blob map[Chunk]struct{})
	dfs = func(chunk Chunk, blob map[Chunk]struct{}) {
		if _, ok := visited[chunk]; ok {
			return
		}
		if _, ok := allLoadedChunks[chunk]; !ok {
			return
		}
		visited[chunk] = struct{}{}
		blob[chunk] = struct{}{}
		for _, neighboring := range chunk.getNeighboring() {
			dfs(neighboring, blob)
		}
	}

	for chunk := range allLoadedChunks {
		if _, ok := visited[chunk]; ok {
			continue
		}
		blob := make(map[Chunk]struct{})
		dfs(chunk, blob)
		blobs = append(blobs, blob)
	}
	cl.blobs = blobs
}

type ServerAssignment struct {
	NumChunks int
	Blobs     []map[Chunk]struct{}
}

func (cl *clusters) redistribute(serverList []Server) map[Server]*ServerAssignment {
	assignments := make(map[Server]*ServerAssignment)
	for _, server := range serverList {
		assignments[server] = new(ServerAssignment)
	}

	for _, blob := range cl.blobs {
		// assign blobs to servers that have the most of the blob already loaded
		serverLoadedCounts := make(map[Server]int)
		var maxServer Server
		var maxCount int
		for chunk := range blob {
			loadedServers := cl.chunkState.getLoadedServers(chunk)
			for server := range loadedServers {
				if _, ok := serverLoadedCounts[server]; !ok {
					serverLoadedCounts[server] = 0
				}
				serverLoadedCounts[server]++
				if serverLoadedCounts[server] > maxCount {
					maxServer = server
					maxCount = serverLoadedCounts[server]
				}
			}
		}
		assignments[maxServer].NumChunks += len(blob)
		assignments[maxServer].Blobs = append(assignments[maxServer].Blobs, blob)
	}
	if len(serverList) == 1 {
		return assignments
	}

	// do rebalancing to handle skew
	for {
		var minChunkServer Server
		minChunkNumber := math.MaxInt64

		var maxChunkServer Server
		maxChunkNumber := math.MinInt64
		var maxChunkLastBlobSize int
		for server, assignment := range assignments {
			if assignment.NumChunks < minChunkNumber {
				minChunkServer = server
				minChunkNumber = assignment.NumChunks
			}
			if assignment.NumChunks > maxChunkNumber {
				if len(assignment.Blobs) == 0 {
					continue
				}
				maxChunkServer = server
				maxChunkNumber = assignment.NumChunks
				maxChunkLastBlobSize = len(assignment.Blobs[len(assignment.Blobs)-1])
			}
		}
		if maxChunkNumber-maxChunkLastBlobSize < minChunkNumber+maxChunkLastBlobSize {
			break
		}
		maxChunkBlobs := assignments[maxChunkServer].Blobs
		var lastBlob map[Chunk]struct{}
		assignments[maxChunkServer].Blobs, lastBlob = maxChunkBlobs[:len(maxChunkBlobs)-1], maxChunkBlobs[len(maxChunkBlobs)-1]
		assignments[minChunkServer].Blobs = append(assignments[minChunkServer].Blobs, lastBlob)
		assignments[maxChunkServer].NumChunks -= maxChunkLastBlobSize
		assignments[minChunkServer].NumChunks += maxChunkLastBlobSize
	}

	// for server, sas := range assignments {
	// 	fmt.Println("num assignment of", server, ":", sas.NumChunks, sas.Blobs)
	// }
	return assignments
}

type Transfer struct {
	From Server
	To   Server
}

func (cl *clusters) GetTransfers(serverList []Server) map[Transfer][]Chunk {
	cl.Lock()
	defer cl.Unlock()
	newAssignments := cl.redistribute(serverList)
	transfers := make(map[Transfer][]Chunk)
	for server, assignment := range newAssignments {
		for _, blob := range assignment.Blobs {
			for chunk := range blob {
				if cl.chunkState.isOwned(chunk) {
					previousOwner := cl.chunkState.getOwner(chunk)
					if previousOwner == server {
						// do nothing, we already own it
					} else {
						// we don't own the chunk, transfer ownership
						key := Transfer{
							From: previousOwner,
							To:   server,
						}
						if _, ok := transfers[key]; !ok {
							transfers[key] = make([]Chunk, 0)
						}
						transfers[key] = append(transfers[key], chunk)
					}
				} else {
					// nobody owns it - we need to set ownership
					cl.chunkState.setOwner(chunk, server)
				}
			}
		}
	}
	return transfers
}

func (cl *clusters) GetServerOwnedChunks(server Server) map[Chunk]struct{} {
	cl.Lock()
	defer cl.Unlock()
	return cl.chunkState.getServerOwnedChunks(server)
}

func (cl *clusters) GetOwner(chunk Chunk) (Server, bool) {
	cl.Lock()
	defer cl.Unlock()
	if !cl.chunkState.isOwned(chunk) {
		return "", false
	}
	return cl.chunkState.getOwner(chunk), true
}

func (cl *clusters) SetOwner(chunk Chunk, owner Server) {
	cl.Lock()
	defer cl.Unlock()
	if cl.chunkState.isOwned(chunk) {
		cl.chunkState.removeOwner(chunk)
	}
	cl.chunkState.setOwner(chunk, owner)
}

func chunkChanges(old map[Chunk]struct{}, current map[Chunk]struct{}) (new map[Chunk]struct{}, unloaded map[Chunk]struct{}) {
	newLoadedChunks := make(map[Chunk]struct{})
	unloadedChunks := make(map[Chunk]struct{})

	for chunk := range current {
		if _, exists := old[chunk]; !exists {
			newLoadedChunks[chunk] = struct{}{}
		}
	}
	for chunk := range old {
		if _, exists := current[chunk]; !exists {
			unloadedChunks[chunk] = struct{}{}
		}
	}
	return newLoadedChunks, unloadedChunks
}
